### DISCUSSION TITLE
Change border colors

### DISCUSSION THREAD

**Discussion by BigfootN** [folke/noice.nvim#778] [Q&A]
Hello,

I would like to be able to change the border colors of the `cmdline_popup` view.

I have tried several highlight groups:
- `NoiceCmdlinePopup`
- `NoiceCmdlinePopupBorder`
- `NoiceCmdlinePopupBorderInput`

Here is my configuration (I'm using `lazy.nvim`):

```lua
	"folke/noice.nvim",
	dependencies = {
		"MunifTanjim/nui.nvim",
	},
	opts = {
		cmdline = {
			format = {
				cmdline = {
					title = "",
				},
			},
			enabled = true,
			view = "cmdline_popup",
		},
	},
	config = function(_, opts)
		border_group_id = vim.api.nvim_get_hl_id_by_name("NoiceCmdlinePopupBorderInput")
		vim.api.nvim_set_hl(border_group_id, "NoiceCmdlinePopupBorderInput", { fg = "#e78a4e" })
		require("noice").setup(opts)
	end
```

Thank you very much in advance for any help

**Comment by Sam-programs** [comment]
```lua
vim.api.nvim_set_hl(border_group_id, "NoiceCmdlinePopupBorderInput", { fg = "#e78a4e" })
```
`nvim_set_hl` takes a namespace id, not a highlight group id. `0` is the global namespace id (which is activated by default).
```md
      • {ns_id}  Namespace id for this highlight |nvim_create_namespace()|.
                 Use 0 to set a highlight group globally |:highlight|.
                 Highlights from non-global namespaces are not active by
                 default, use |nvim_set_hl_ns()| or |nvim_win_set_hl_ns()| to
                 activate them.
```

**Reply by BigfootN** [reply]
Thank you for your quick reply.

Unfortunately I tried to set it to `0`, but the result was the same.

**Reply by Sam-programs** [reply]
For each format in `config.cmdline.format`, Noice uses the highlight `NoiceCmdlinePopupBorder<Captalized name of format>` for the cmdline border (e.g. the `cmdline` format would use the highlight `NoiceCmdlinePopupBorderCmdline`)
By default, most formats highlights are linked to `NoiceCmdlinePopupBorder`, except search which uses `NoiceCmdlinePopupBorderSearch`.
So, in short, you would need to set both `NoiceCmdlinePopupBorder` and `NoiceCmdlinePopupBorderSearch` to change the border for every format.

**Reply by BigfootN** [reply]
Thank you for your time.

Unfortunately it didn't solve the problem.

This is what I have as a the `config` function:

```lua
		vim.api.nvim_set_hl(0, "NoiceCmdlinePopupBorderCmdLine", { fg = "#473c29" })
		vim.api.nvim_set_hl(0, "NoiceCmdlinePopupBorderSearch", { fg = "#473c29" })
		vim.api.nvim_set_hl(0, "NoiceCmdlinePopupBorder", { fg = "#473c29" })
		require("noice").setup(opts)
		vim.api.nvim_set_hl(0, "NoiceCmdlinePopupBorderCmdLine", { fg = "#473c29" })
		vim.api.nvim_set_hl(0, "NoiceCmdlinePopupBorderSearch", { fg = "#473c29" })
		vim.api.nvim_set_hl(0, "NoiceCmdlinePopupBorder", { fg = "#473c29" })
```

What is _**really weird**_ is that it seamed to work for a certain time, but the origin color came back. It seams "random".

N.B.: I set the highlight before and after the `setup` just to make sure it was _**actually set**_ 

Thank you for your time and help.

**Reply by Sam-programs** [reply]
> What is really weird is that it seamed to work for a certain time, but the origin color came back. It seams "random".

You can use `nvim -V1logfile` to enable verbose logging, then when you use the `:highlight <highlight>` neovim will write the file where  `<highlight>` was last changed into `logfile` 
(e.g. using `:highlight NoiceCmdlinePopupBorder` will write something like this: 
```
NoiceCmdlinePopupBorderxxx guifg=<some color>
	Last set from <file where it was last changed> line <line number>
```
into `logfile`
After the origin color comes back, you can check `logfile` and see what changed the highlights.

**Reply by BigfootN** [reply]
Thank you for your great answer and time.

I did exactly that and here is what I got:

```
NoiceCmdlinePopupBorderCmdLinexxx guifg=#473c29
	Last set from ~/.config/nvim/lua/plugins/ui.lua line 24


NoiceCmdlinePopupBorderCmdLinexxx links to NoiceCmdlinePopupBorder
	Last set from ~/.local/share/nvim/lazy/noice.nvim/lua/noice/config/highlights.lua line 104
```

The two seem completely inconsistent. One talks about the color set from `noice.vim` and the other says it was "well set" from my configuration.

**Reply by Sam-programs** [reply]
> `	Last set from ~/.local/share/nvim/lazy/noice.nvim/lua/noice/config/highlights.lua line 104`

The code here doesn't define a highlight; unless, it is cleared. I am guessing the theme changed and cleared the highlights. Maybe define the highlights in a `ColorScheme` autocmd:
```lua
vim.api.nvim_set_hl(0, "NoiceCmdlinePopupBorder", { fg = "#473c29" })
vim.api.nvim_set_hl(0, "NoiceCmdlinePopupBorderCmdLine", { fg = "#473c29" })
vim.api.nvim_set_hl(0, "NoiceCmdlinePopupBorderSearch", { fg = "#473c29" })
vim.api.nvim_create_autocmd("ColorScheme", {
    callback = function()
        vim.api.nvim_set_hl(0, "NoiceCmdlinePopupBorder", { fg = "#473c29" })
        vim.api.nvim_set_hl(0, "NoiceCmdlinePopupBorderCmdLine", { fg = "#473c29" })
        vim.api.nvim_set_hl(0, "NoiceCmdlinePopupBorderSearch", { fg = "#473c29" })
    end
})
require('noice').setup(opt)
```
or when your theme loads.

**Reply by BigfootN** [reply]
Thank you for your answer.

I think I found a "better"/"simpler" solution. To be honnest I haven't tested yours.

In my "solution" I just added `force = true` to the options of `nvim_set_hl`.

Thank you for your time.
